pkg/datapath/linux/route/reconciler/table.go

```diff
@@ -0,0 +1,364 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package reconciler
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net/netip"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/statedb"
+	"github.com/cilium/statedb/index"
+	"github.com/cilium/statedb/reconciler"
+
+	"github.com/cilium/cilium/pkg/datapath/tables"
+)
+
+var _ statedb.TableWritable = &DesiredRoute{}
+
+type DesiredRouteKey struct {
+	Owner    *RouteOwner
+	Table    TableID
+	Prefix   netip.Prefix
+	Priority uint32
+}
+
+func (k DesiredRouteKey) Key() index.Key {
+	var key []byte
+	if k.Owner != nil {
+		key = index.String(k.Owner.name)
+		// Owner name is variable, insert a 0x00 byte to separate it from the rest of the key.
+		key = append(key, 0x00)
+	}
+
+	if k.Table != 0 {
+		key = append(key, index.Uint32(uint32(k.Table))...)
+	}
+
+	if k.Prefix.IsValid() {
+		key = append(key, index.NetIPAddr(k.Prefix.Addr())...)
+		key = append(key, uint8(k.Prefix.Bits()))
+	}
+
+	if k.Priority != 0 {
+		key = append(key, index.Uint32(k.Priority)...)
+	}
+
+	return key
+}
+
+func (k DesiredRouteKey) String() string {
+	parts := []string{k.Owner.String(), k.Table.String(), k.Prefix.String()}
+	if k.Priority != 0 {
+		parts = append(parts, strconv.FormatUint(uint64(k.Priority), 10))
+	}
+	return strings.Join(parts, ":")
+}
+
+func (k DesiredRouteKey) MarshalBinary() ([]byte, error) {
+	var (
+		buf []byte
+		tmp [4]byte
+	)
+
+	binary.LittleEndian.PutUint32(tmp[:], uint32(k.Table))
+	buf = append(buf, tmp[:]...)
+
+	addrBuf, err := k.Prefix.Addr().MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	buf = append(buf, byte(len(addrBuf)))
+	buf = append(buf, addrBuf...)
+	buf = append(buf, byte(k.Prefix.Bits()))
+
+	binary.LittleEndian.PutUint32(tmp[:], k.Priority)
+	buf = append(buf, tmp[:]...)
+
+	return buf, nil
```

Comment:
This binary will be persisted to the WAL right? Then I'd suggest encoding field identifier (like protobuf) or format version identifier somewhere. Otherwise, upgrade/downgrade would be painful whenever we change the key structure.

Reply:
I guess a version number couldn't hurt. I hope to never change it since its the primary key of a route on the routing table, which is sort of static. But having that number if ever something unforeseen happens might save someone a lot of pain.